nav2_map_server/src/map_saver_cli.cpp

```diff
@@ -0,0 +1,108 @@
+// Copyright 2019 Rover Robotics
+// Copyright (c) 2008, Willow Garage, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "nav2_map_server/map_saver.hpp"
+#include "rclcpp/rclcpp.hpp"
+
+const char * USAGE_STRING{
+  "Usage: \n"
+  "  map_saver -h/--help\n"
+  "  map_saver [--occ <threshold_occupied>] [--free <threshold_free>] [--fmt <image_format>] "
+  "[--mode trinary/scale/raw] [-f <mapname>] [ROS remapping args]"};
+
+int main(int argc, char ** argv)
+{
+  rclcpp::init(argc, argv);
+
+  auto logger = rclcpp::get_logger("map_saver_cli");
+
+  std::vector<std::string> arguments(argv + 1, argv + argc);
+  std::vector<rclcpp::Parameter> params_from_args;
+  for (auto it = arguments.begin(); it != arguments.end(); it++) {
+    if (*it == "-h" || *it == "--help") {
+      std::cout << USAGE_STRING << std::endl;
+      return 0;
+    } else if (*it == "-f") {
+      if (++it == arguments.end()) {
+        RCLCPP_WARN(logger, "Argument ignored: -f should be followed by a value.");
+        continue;
+      }
+      params_from_args.emplace_back("output_file_no_ext", *it);
+    } else if (*it == "--occ") {
+      if (++it == arguments.end()) {
+        RCLCPP_WARN(logger, "Argument ignored: --occ should be followed by a value.");
+        continue;
+      }
+      params_from_args.emplace_back("threshold_occupied", atoi(it->c_str()));
+    } else if (*it == "--free") {
+      if (++it == arguments.end()) {
+        RCLCPP_WARN(logger, "Argument ignored: --free should be followed by a value.");
+        continue;
+      }
+      params_from_args.emplace_back("threshold_free", atoi(it->c_str()));
+    } else if (*it == "--mode") {
+      if (++it == arguments.end()) {
+        RCLCPP_WARN(logger, "Argument ignored: --mode should be followed by a value.");
+        continue;
+      }
+      params_from_args.emplace_back("map_mode", *it);
+    } else if (*it == "--fmt") {
+      if (++it == arguments.end()) {
+        RCLCPP_WARN(logger, "Argument ignored: --fmt should be followed by a value.");
+        continue;
+      }
+      params_from_args.emplace_back("image_format", *it);
+    } else {
+      RCLCPP_WARN(logger, "Ignoring unrecognized argument '%s'", it->c_str());
+    }
+  }
+
+  auto node = std::make_shared<nav2_map_server::MapSaver>(
```

Comment:
This feels over complicated, I would think the first thing you do in main() is to make a node so you don't have to create a separate logger, and then hand that node over to the library of `MapSaver` in the constructor, then you can just make a call to `bool map_saver.save_map(args)` for success/failure and have it use that `node` to log internally, there's some `cout` things in the `spin_*`. It feels unnecessary to really involve `rclcpp` behind logging & creating the sub subscription. 

The map saver that this CLI wraps I don't believe really needs to deal with futures.

Reply:
This was a considered choice. The reason I did it this way is so it uses parameters like any other node and composes nicely in a launch file. Argument handling is for convenience and backwards compatibility.

I would prefer to keep argument handling and logging outside the node.

Reply:
Also, the reason for doing futures is so you can request the map multiple times if you so choose.

Reply:
I feel like then there should be a server that is just running that someone is querying to save, using a shared library between that and the generic single-off saver.

The reason is that it's generally good form when possible to wrap a ROS-less lib in ROS stuff, and the actual map_saver, I don't actually see any reason to have it access ROS at all. You could conceivably create the Node, a callback that hands the message to the library to parse and save without giving it the node at all and use exceptions/error codes/bools to give info back to the logger in the `cli` node. Its cleaner to read (IMO) and keeps an independent library from ROS, that could be also used to make a server version that can be used to continuously save maps / called.

Reply:
It accesses ROS because it needs ROS's DDS implementation.

Reply:
Can you explain? The process of Occ map -> file doesn’t seem to me to require ROS at all.

Reply:
MapSaver requires listening for a map using ROS.
There is also some map manipulation code in nav2_util https://github.com/ros-planning/navigation2/tree/master/nav2_util/ which *does* provide an attempt at an API. Extracting this to an API should be done and will not be done in this PR.

Reply:
I think we're missing each other, but that's OK. I'm not blocking over this. I'm thinking something like

lib:
```
class SaverThing
{
  SaverThing(params);
  SomeReturnEnumType SaveThis(occ_grid::Ptr map)
  {
    // save map as file in params  
  }
}

```

executable:
```
SaverThing * obj;

void mapCb(map) {obj->saveMap(map);}

main()
{
  node = NodeDef()
  obj = new SaverThing(params_taken_from_node)
  node->create_subscription...
  return 1;

}
```

Give or take. Then all the ROS stuff is in the executible and the library is clean of just taking the ROS message type of a map and converting it into an image file of whatever parameterization. This buys you a library that can:
- Be used to save a map once, with a simple wrapper
- Be used to save N maps N times, with a service wrapper
- Be used in another more complex application that can #include<nav2_map_server/lib.h> and then save some other occ grid type thing as it likes (ei some weird costmap stuff or what have you

Decoupling ROS entirely buys you more re-usability in my head. Making the library as a Node seems to me overly restrictive

Reply:
You're right, that it's a nice to have and I'd really like to do that when I reconcile this code with loadMapFromFile from nav2_util. Can you file that as an issue to get back to later?

Additionally, I did not yet make sense of map_loader.hpp, which seems to be some sort of interface to enable other classes like OccGridLoader. I'm not sure the future of that file nor the other map formats we intend to support, and an API design here should take that into account.

Reply:
Sure.

That was an early attempt to add other map types, but we never really got there, and I would agree with you its extremely hard to make sense of. I would just ignore it and write it as you would want to have it written and nuke parts of it as needed. We can circle back to supporting other map types when its actually relevant and someone's willing to put in the time to fully support. 

I'd even go a step further and move that code back into `map_server'. There's even an old-school [REP](http://www.ros.org/reps/rep-0144.html) for not having utils packages and I've been slowly trying to get it organized in a way that it belongs as some package.

Reply:
That sounds great. I'll start on another PR to clean up this mess a bit more.